diff --git a/pyproject.toml b/pyproject.toml
index 7bc67441e5..55c73cb255 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -71,18 +71,8 @@ script-files = [
   "bin/igwn_ligolw_sqlite",
 ]
 
-[[tool.setuptools.ext-modules]]
-name = "igwn_ligolw.tokenizer"
-sources = [
-  "src/igwn_ligolw/tokenizer.c",
-  "src/igwn_ligolw/tokenizer.Tokenizer.c",
-  "src/igwn_ligolw/tokenizer.RowBuilder.c",
-  "src/igwn_ligolw/tokenizer.RowDumper.c"
-]
-include-dirs = ["src/igwn_ligolw"]
-
 [tool.setuptools_scm]
-version_file = "src/igwn_ligolw/_version.py"
+write_to = "src/igwn_ligolw/_version.py"
 
 [tool.coverage.paths]
 source = [
diff --git a/setup.py b/setup.py
new file mode 100644
index 0000000000..a1a277acaa
--- /dev/null
+++ b/setup.py
@@ -0,0 +1,16 @@
+from setuptools import Extension, setup
+
+setup(
+    ext_modules=[
+        Extension(
+            "igwn_ligolw.tokenizer",
+            [
+                "src/igwn_ligolw/tokenizer.c",
+                "src/igwn_ligolw/tokenizer.Tokenizer.c",
+                "src/igwn_ligolw/tokenizer.RowBuilder.c",
+                "src/igwn_ligolw/tokenizer.RowDumper.c",
+            ],
+            include_dirs=["src/igwn_ligolw"],
+        ),
+    ],
+)
