Index: src/pax1000_controller/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/pax1000_controller/__init__.py b/src/pax1000_controller/__init__.py
new file mode 100644
--- /dev/null	(date 1768229815907)
+++ b/src/pax1000_controller/__init__.py	(date 1768229815907)
@@ -0,0 +1,1 @@
+from .pax1000_controller import *
\ No newline at end of file
diff --git a/pax1000_controller.py b/src/pax1000_controller/pax1000_controller.py
rename from pax1000_controller.py
rename to src/pax1000_controller/pax1000_controller.py
